src/ops/indexing.rs

```diff
@@ -0,0 +1,1799 @@
+                strides.len(),
+                stream.as_ptr(),
+            );
+
+            Array::from_ptr(c_array)
+        }
+    }
+
+    pub(crate) fn try_slice_device(
+        &self,
+        start: &[i32],
+        stop: &[i32],
+        strides: &[i32],
+        stream: StreamOrDevice,
+    ) -> Result<Array, SliceError> {
+        if start.len() != self.ndim() || stop.len() != self.ndim() || strides.len() != self.ndim() {
+            return Err(SliceError { ndim: self.ndim() });
+        }
+
+        unsafe { Ok(self.slice_device_unchecked(start, stop, strides, stream)) }
+    }
+
+    pub(crate) fn slice_device(
+        &self,
+        start: &[i32],
+        stop: &[i32],
+        strides: &[i32],
+        stream: StreamOrDevice,
+    ) -> Array {
+        self.try_slice_device(start, stop, strides, stream).unwrap()
+    }
+}
+
+/* -------------------------------------------------------------------------- */
+/*                              Helper functions                              */
+/* -------------------------------------------------------------------------- */
+
+fn absolute_indices(absolute_start: i32, absolute_end: i32, stride: i32) -> Vec<i32> {
+    let mut indices = Vec::new();
+    let mut i = absolute_start;
+    while (stride > 0 && i < absolute_end) || (stride < 0 && i > absolute_end) {
+        indices.push(i);
+        i += stride;
+    }
+    indices
+}
+
+// Implement additional public APIs
+//
+// TODO: rewrite this in a more rusty way
+#[inline]
+fn gather_nd<'a>(
+    src: &'a Array,
+    operations: impl Iterator<Item = &'a ArrayIndexOp>,
+    gather_first: bool,
+    stream: StreamOrDevice,
+) -> (usize, Array) {
+    use ArrayIndexOp::*;
+
+    let mut max_dims = 0;
+    let mut slice_count = 0;
+    let mut is_slice: Vec<bool> = Vec::new();
+    let mut gather_indices: Vec<Rc<Array>> = Vec::new();
+
+    let shape = src.shape();
+
+    // prepare the gather indices
+    let mut axes = Vec::new();
+    let mut operation_len: usize = 0;
+    for (i, op) in operations.enumerate() {
+        axes.push(i as i32);
+        operation_len += 1;
+        match op {
+            TakeIndex { index } => {
+                let item = Array::from_int(resolve_index_unchecked(
+                    *index,
+                    src.dim(i as i32) as usize,
+                ) as i32);
+                gather_indices.push(Rc::new(item));
+                is_slice.push(false);
+            }
+            Slice(range) => {
+                slice_count += 1;
+                is_slice.push(true);
+
+                let size = shape[i];
+                let absolute_start = range.absolute_start(size);
+                let absolute_end = range.absolute_end(size);
+
+                // // TODO: check if this is correct when stride is negative
+                // let indices: Vec<i32> = (absolute_start..absolute_end)
+                //     .step_by(range.stride().abs() as usize)
+                //     .collect();
```

Comment:
is this meant to be resolved or unit tested?

Reply:
That was already replaced by the absolute_indices line below.

Reply:
Addressed in 55ca17d931a3dff70b189ecb3086957543ca4d55